### ISSUE TITLE
[M3] Freeze held-out benchmark workload suite

### ISSUE THREAD

**Issue by wisoums** [wisoums/scalerl#18]
Freeze ScaleRL's train/validation/held-out workload suite **before DQN/PPO tuning**.

This issue creates the versioned benchmark manifest and the workload-construction API used by later training/evaluation code.

The final test set must never be used for controller hyperparameter selection, reward tuning, early stopping, model selection, or simulator-capacity tuning.

## Benchmark version

Create a versioned benchmark directory, preferably:

```text
benchmarks/v1/
├── workloads.json
└── README.md
```

Use JSON rather than introducing a YAML dependency.

The manifest is committed and immutable in meaning once v1 results begin. Any later change to held-out membership requires a new benchmark version or an explicit documented correction.

## Common episode contract

For the main v1 comparison, every manifest workload uses:

- `control_interval_seconds = 30.0`
- `duration_seconds = 3600.0`

This keeps:
- episode length identical;
- the core simulator configuration compatible;
- learned-policy observation/action spaces comparable.

Workload demand itself may differ; do not normalize Azure demand to match synthetic demand.

## Split semantics

Every workload has exactly one split:

- `train` — policy learning only;
- `validation` — threshold/predictive/RL tuning and model selection;
- `test` — final reporting only.

The split is public/versioned but **held out by process**: training/tuning code must not consume `test` workloads.

## Synthetic workloads

Freeze the following v1 synthetic scenarios exactly.

All use 3600 s duration and 30 s control interval.

### Train

1. `syn-train-steady-moderate`
   - generator: `steady_workload`
   - `rate = 100.0`

2. `syn-train-seasonal`
   - generator: `diurnal_workload`
   - `baseline = 150.0`
   - `amplitude = 100.0`
   - `period_seconds = 3600.0`
   - `phase_seconds = 0.0`

3. `syn-train-ramp-up`
   - generator: `ramp_workload`
   - `start_rate = 40.0`
   - `end_rate = 360.0`

4. `syn-train-spike`
   - generator: `spike_workload`
   - `baseline = 80.0`
   - `spike_rate = 420.0`
   - `spike_start_seconds = 1500.0`
   - `spike_duration_seconds = 300.0`

5. `syn-train-bursty`
   - generator: `bursty_workload`
   - `baseline = 90.0`
   - `burst_rate = 250.0`
   - `burst_probability = 0.08`
   - `burst_duration_seconds = 90.0`
   - `noise_std = 10.0`
   - `seed = 101`

### Validation

1. `syn-val-steady-high`
   - steady
   - `rate = 250.0`

2. `syn-val-ramp-down`
   - ramp
   - `start_rate = 350.0`
   - `end_rate = 60.0`

3. `syn-val-bursty`
   - bursty
   - `baseline = 120.0`
   - `burst_rate = 280.0`
   - `burst_probability = 0.12`
   - `burst_duration_seconds = 60.0`
   - `noise_std = 15.0`
   - `seed = 202`

### Held-out test

1. `syn-test-seasonal-shifted`
   - diurnal/seasonal
   - `baseline = 180.0`
   - `amplitude = 140.0`
   - `period_seconds = 3600.0`
   - `phase_seconds = 900.0`

2. `syn-test-spike-hard`
   - spike
   - `baseline = 60.0`
   - `spike_rate = 480.0`
   - `spike_start_seconds = 900.0`
   - `spike_duration_seconds = 180.0`

These values are frozen before deep-RL tuning. Do not silently retune them after observing model performance.

## Azure workloads

Use the Azure Functions Invocation Trace 2021 loader from #43.

The official trace spans two weeks and the dataset-specific documentation defines timestamps/durations in seconds; timestamps are modified, so all selections are trace-relative.

Use exact fixed offsets rather than visually choosing "interesting" windows after looking at controller results.

All Azure slices:
- `duration_seconds = 3600.0`
- `control_interval_seconds = 30.0`

Freeze these offsets:

### Train

- `azure-train-129600`: `start_seconds = 129600.0`
- `azure-train-302400`: `start_seconds = 302400.0`
- `azure-train-475200`: `start_seconds = 475200.0`

### Validation

- `azure-val-734400`: `start_seconds = 734400.0`

### Held-out test

- `azure-test-993600`: `start_seconds = 993600.0`
- `azure-test-1166400`: `start_seconds = 1166400.0`

These are deliberately spread across the two-week trace rather than selected based on observed controller performance.

If a frozen Azure window later proves unusable because it is outside the actual extracted file, empty due to source-data issues, or corrupted, do **not** silently substitute another window. Document the reason and make an explicit benchmark-version correction before any final model evaluation.

## Manifest schema

Each manifest entry should include at least:

```json
{
  "id": "...",
  "split": "train | validation | test",
  "source": "synthetic | azure",
  "duration_seconds": 3600.0,
  "control_interval_seconds": 30.0,
  "generator": "...",
  "parameters": {}
}
```

For Azure entries, store:
- logical dataset identifier;
- exact `start_seconds`;
- duration;
- interval;

but **not** an absolute local filesystem path.

The raw Azure CSV path is supplied at runtime by the caller/CLI.

## Manifest API

Add a small typed benchmark module/package that:

- loads and validates the committed v1 JSON manifest;
- exposes entries by split;
- constructs synthetic `WorkloadTrace` objects directly;
- constructs Azure `WorkloadTrace` objects when given an Azure CSV path;
- preserves the existing `WorkloadTrace` abstraction;
- does not introduce controller-specific behavior.

A conceptual API is acceptable:

```python
manifest = load_benchmark_manifest(...)
manifest.train
manifest.validation
manifest.test

trace = build_workload(entry, azure_csv_path=...)
```

Exact names should follow project style.

Use Pydantic or existing project validation conventions; do not add a new schema dependency.

## Guardrails

Make accidental test-set use harder and explicit.

At minimum:
- split is a validated enum/literal;
- IDs are unique;
- every entry belongs to exactly one split;
- training/validation selection helpers do not silently include test workloads;
- no "all workloads for tuning" helper that casually merges test with train/validation.

Do not try to create a security boundary; this is an experiment-integrity contract.

## Azure data validation / characterization

The full Azure dataset is not available in CI.

Provide a small local-only command/script or module function that, when given the extracted Azure CSV:
- loads all Azure manifest entries;
- verifies each window is non-empty and valid;
- reports deterministic workload characteristics such as:
  - mean RPS;
  - max RPS;
  - standard deviation;
  - coefficient of variation when defined;
  - peak/mean ratio when defined;
- optionally writes a small committed-safe summary JSON containing only aggregate statistics/provenance, never raw events.

Do **not** use these statistics to change the already-frozen test membership.

If the raw Azure dataset is not present during implementation, do not invent Azure statistics. The structural manifest/tests can still be implemented, but #18 should only be considered fully complete after the selected Azure windows have been locally validated against the real extracted trace.

## Benchmark documentation

`benchmarks/v1/README.md` should document:

- benchmark version and freeze date;
- why train/validation/test are separate;
- exact held-out rule;
- synthetic scenario intent/difficulty;
- Azure fixed-offset selection rule;
- no normalization of Azure demand;
- required local Azure data setup from `data/README.md`;
- any committed aggregate characterization after local validation;
- rule that final reported results use test only after model/hyperparameter choices are frozen.

## Simulator compatibility

Do not embed a new simulator implementation into the benchmark layer.

For the main v1 comparison:
- all workloads have 30 s interval and 3600 s duration;
- controllers share the same simulator configuration per comparison;
- learned policies must use compatible observation/action spaces;
- simulator-capacity choices are experiment configuration, not workload normalization.

Do not calibrate service capacity using held-out test workload statistics.

## Tests

Tests must be offline and use synthetic workloads plus the tiny Azure fixture from #43.

Cover at least:

### Manifest validation
- manifest loads successfully;
- workload IDs are unique;
- invalid split/source/generator rejected;
- missing required parameters rejected;
- duration/interval mismatch rejected;
- every v1 workload is exactly 120 ticks.

### Split integrity
- expected train IDs are only in train;
- expected validation IDs are only in validation;
- held-out IDs are only in test;
- helpers for train/validation cannot accidentally return test workloads.

### Synthetic reproducibility
- every synthetic manifest entry reconstructs deterministically;
- bursty seeds are frozen and reproducible.

### Azure spec
- exact Azure offsets above are preserved;
- Azure windows do not overlap one another;
- runtime source path is external to the manifest;
- tiny #43 fixture can be wired through the benchmark builder.

### Integration
- at least one synthetic benchmark entry and fixture-backed Azure benchmark entry can create `AutoscalingEnv`;
- existing `run_episode` works without benchmark-specific controller logic.

## Scope

Do **not** implement:
- MLflow tracking (#17);
- threshold grid tuning (#13);
- predictive controller (#14);
- Docker (#44);
- expanded CI (#45);
- Q-learning/DQN/PPO;
- multi-seed statistics (#19);
- final benchmark results (#46);
- any real test-set model evaluation.

This issue freezes data membership and provides reproducible workload construction only.

## Acceptance criteria

- [ ] committed versioned v1 benchmark manifest exists
- [ ] train/validation/test membership is explicit and immutable in meaning
- [ ] synthetic parameters above are frozen and reproducible
- [ ] Azure offsets above are frozen and reproducible
- [ ] all workloads use 3600 s / 30 s common episode contract
- [ ] test workloads cannot be included by train/validation helpers accidentally
- [ ] benchmark loader builds existing `WorkloadTrace` objects
- [ ] Azure source path stays runtime/local, not committed in manifest
- [ ] local Azure validation/characterization path exists
- [ ] no held-out stats are used to tune simulator/controller/model settings
- [ ] benchmark README documents freeze/leakage rules
- [ ] offline tests cover manifest validation, split integrity, reproducibility, and integration
- [ ] Ruff, Ruff format check, mypy, and pytest pass
